src/pip/_internal/req/__init__.py

```diff
@@ -53,40 +60,117 @@ def install_given_reqs(
             ', '.join([req.name for req in to_install]),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = [None] * len(
+        to_install
+    )  # type: List[Union[None, InstallationResult, BaseException]]
+
+    install_args = [install_options, global_options, args, kwargs]
 
     with indent_log():
-        for requirement in to_install:
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', requirement.name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    *args,
-                    **kwargs
-                )
-            except Exception:
-                should_rollback = (
-                    requirement.should_reinstall and
-                    not requirement.install_succeeded
-                )
-                # if install did not succeed, rollback previous uninstall
-                if should_rollback:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                should_commit = (
-                    requirement.should_reinstall and
-                    requirement.install_succeeded
-                )
-                if should_commit:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(requirement.name))
-
-    return installed
+        # first try to install in parallel
+        installed_pool = __safe_pool_map(
+            partial(__single_install, install_args, in_subprocess=True),
+            to_install)
+        if installed_pool:
+            installed = installed_pool
+
+        for i, requirement in enumerate(to_install):
+            if installed[i] is None:
+                installed[i] = __single_install(
+                    install_args, requirement, in_subprocess=False)
+            elif isinstance(installed[i], BaseException):
+                raise installed[i]   # type: ignore
+
+    return [i for i in installed if isinstance(i, InstallationResult)]
+
+
+def __safe_pool_map(
+        func,               # type: Callable[[Any], Any]
+        iterable,           # type: Iterable[Any]
+):
+    # type: (...) -> Optional[List[Any]]
+    """
+    Safe call to Pool map, if Pool is not available return None
+
+    """
+    # Disable multiprocessing on Windows python 2.7
+    if sys.platform == 'win32' and sys.version_info.major == 2:
+        return None
+
+    if not iterable or Pool is None:
+        return None
+
+    # first let's try to install in parallel,
+    # if we fail we do it by order.
+    pool = None
+    results = None
+    try:
+        pool = Pool()
```

Comment:
If we use Pool using context manager `with pool as Pool()`, I think we will be able to avoid the explicity `pool.close` and `pool.join` calls below

Reply:
Python 2.7 is a bummer, Pool context is not supported :(

Reply:
Aah yes you are right. But I guess we can add a comment about why we are not using pool context (because of python 2.7) and when Python 2 support in removed in future, that comment will help refactor the code to use Pool context.

Reply:
Yes, that's a good idea. will do.